<a href="https://www.bigdatauniversity.com"><img src = "https://ibm.box.com/shared/static/kpwal4mklabkeipypfo47rrm8xq0lslk.jpg" width = 300, align = "center"></a>

<h1 align=center><font size = 5>Python BASICS</font></h1> 

### Welcome!

By the end of this notebook, you will have learned the basics of Python!  

## Table of Contents


<div class="alert alert-block alert-info" style="margin-top: 20px">
<li><a href="#ref0">About the Dataset</a></li>
<li><a href="#ref1">Simple Math in Python</a></li>
<li><a href="#ref2">Variables in Python</a></li>
<li><a href="#ref1a">Quiz on Simple Math and Variables in Python</a></li>
<li><a href="#ref9">Strings in Python</a></li>
<li><a href="#ref10">Quiz on Strings</a></li>

<br>
<p></p>
Estimated Time Needed: <strong>25 min</strong>
</div>

<hr>

<a id="ref0"></a>
<h2 align=center>About the Dataset</h2>

Which album should you listen to next? 

Let's say each of your friends tells you their favorite top-selling album. You do some research on the albums and put it all into a table. Now you can begin exploring the dataset, and asking questions about the albums and artists. For example, you can check if albums from some certain genres tend to get better ratings. You can check how the sales for albums changes across years, and much more. 

**Movies dataset**

The table gathered includes one row for each movie, with several columns for each album characteristic:

- **artist** - Name of the artist
- **album** - Name of the album
- **released_year** - Year the album was released
- **length_min_sec** - Length of the album (hours,minutes,seconds)
- **genre** - Genre of the album
- **music_recording_sales_millions** - Music recording sales (millions in USD) on [SONG://DATABASE](http://www.song-database.com/)
- **claimed_sales_millions** - Album's claimed sales (millions in USD) on [SONG://DATABASE](http://www.song-database.com/)
- **date_released** - Date on which the album was released
- **soundtrack** - Indicates if the album is the movie soundtrack (Y) or (N)
- **rating_of_friends** - Indicates the rating from your friends from 1 to 10
<br>


<font size="1">
<table font-size:xx-small style="width:25%">
  <tr>
    <th>Artist</th>
    <th>Album</th> 
    <th>Released</th>
    <th>Length</th>
    <th>Genre</th> 
    <th>Music recording sales (millions)</th>
    <th>Claimed sales (millions)</th>
    <th>Released</th>
    <th>Soundtrack</th>
    <th>Rating (friends)</th>
  </tr>
  <tr>
    <td>Michael Jackson</td>
    <td>Thriller</td> 
    <td>1982</td>
    <td>00:42:19</td>
    <td>Pop, rock, R&B</td>
    <td>46</td>
    <td>65</td>
    <td>30-Nov-82</td>
    <td></td>
    <td>10.0</td>
  </tr>
  <tr>
    <td>AC/DC</td>
    <td>Back in Black</td> 
    <td>1980</td>
    <td>00:42:11</td>
    <td>Hard rock</td>
    <td>26.1</td>
    <td>50</td>
    <td>25-Jul-80</td>
    <td></td>
    <td>8.5</td>
  </tr>
    <tr>
    <td>Pink Floyd</td>
    <td>The Dark Side of the Moon</td> 
    <td>1973</td>
    <td>00:42:49</td>
    <td>Progressive rock</td>
    <td>24.2</td>
    <td>45</td>
    <td>01-Mar-73</td>
    <td></td>
    <td>9.5</td>
  </tr>
    <tr>
    <td>Whitney Houston</td>
    <td>The Bodyguard</td> 
    <td>1992</td>
    <td>00:57:44</td>
    <td>Soundtrack/R&B, soul, pop</td>
    <td>26.1</td>
    <td>50</td>
    <td>25-Jul-80</td>
    <td>Y</td>
    <td>7.0</td>
  </tr>
    <tr>
    <td>Meat Loaf</td>
    <td>Bat Out of Hell</td> 
    <td>1977</td>
    <td>00:46:33</td>
    <td>Hard rock, progressive rock</td>
    <td>20.6</td>
    <td>43</td>
    <td>21-Oct-77</td>
    <td></td>
    <td>7.0</td>
  </tr>
    <tr>
    <td>Eagles</td>
    <td>Their Greatest Hits (1971-1975)</td> 
    <td>1976</td>
    <td>00:43:08</td>
    <td>Rock, soft rock, folk rock</td>
    <td>32.2</td>
    <td>42</td>
    <td>17-Feb-76</td>
    <td></td>
    <td>9.5</td>
  </tr>
    <tr>
    <td>Bee Gees</td>
    <td>Saturday Night Fever</td> 
    <td>1977</td>
    <td>1:15:54</td>
    <td>Disco</td>
    <td>20.6</td>
    <td>40</td>
    <td>15-Nov-77</td>
    <td>Y</td>
    <td>9.0</td>
  </tr>
    <tr>
    <td>Fleetwood Mac</td>
    <td>Rumours</td> 
    <td>1977</td>
    <td>00:40:01</td>
    <td>Soft rock</td>
    <td>27.9</td>
    <td>40</td>
    <td>04-Feb-77</td>
    <td></td>
    <td>9.5</td>
  </tr>
</table></font>
    

### We can use Python to help us explore the dataset
But to begin, we'll need to start from the basics, so let's get started!

<hr>

<a id="ref1"></a>
<h2 align=center> Simple Math in Python </h2>

Let's say you want to listen to *Thriller*, *Saturday Night Fever*, and *Rumours* back-to-back. Do you have enough time to **listen to all three albums in 2 hours?** Let's try using simple math in Python.  

What is the **total album length** (in minutes; rounding up the seconds so that you don't miss a single note) for Thriller, Saturday Night Fever, and Rumers?
- **Thriller**: 42 min 19 sec
- **Saturday Night Fever**: 1 hour 15 min 54 sec
- **Rumours**: 40 min 1 sec

<div class="alert alert-success alertsuccess" style="margin-top: 20px">
**Tip**: To run the grey code cell below, click on it, and press Shift + Enter.
</div>

In [1]:
43 + 60 + 16 + 41

160

Great! You've determined that the total number of album play time is **160 min**.  

**What is 160 min in hours?**

In [2]:
160.0 / 60.0

2.6666666666666665

Well, it looks like it's well **over 2 hours**, which means you can't listen to *Thriller*, *Saturday Night Fever* and *Rumours* back-to-back without missing out on some of the songs!

Album 1 is 2 hours, album 2 is 30 minutes what is the total length of both albums in minutes.    

In [3]:
 2*60+30

150

<a ><img src = "https://ibm.box.com/shared/static/cc7z5xbg4sorawqye2izp4douptopc1x.gif" width = 500, align = "center"></a>

<div class="alert alert-success alertsuccess" style="margin-top: 20px">
<h4> [Tip] Simple math in Python </h4>
<p></p>
You can do a variety of mathematical operations in Python including:  
<li> addition: **2 + 2** </li>
<li> subtraction: **5 - 2** </li>
<li> multiplication: **3 \* 2** </li>
<li> division: **4 / 2** </li>
<li> exponentiation: **4 \*\* 2** </li>
</div>

<a id="ref2"></a>
<h2 align=center> Variables in Python </h2>

We can also **store** our output in **variables**, so we can use them later on. For example:

In [4]:
x = 43 + 60 + 16 + 41

To return the value of **`x`**, we can simply run the variable as a command:

In [6]:
print(x)

160


We can also perform operations on **`x`** and save the result to a **new variable**:

In [7]:
y = x / 60
y

2.6666666666666665

If we save something to an **existing variable**, it will **overwrite** the previous value:

In [8]:
x = x / 60
x

2.6666666666666665

It's good practice to use **meaningful variable names**, so you don't have to keep track of what variable is what:

In [9]:
total_min = 43 + 60 + 16 + 41
total_min

160

In [10]:
total_hr = total_min / 60  
total_hr

2.6666666666666665

You can put this all into a single expression, but remember to use **round brackets** to add together the album lengths first, before dividing by 60.

In [11]:
total_hr = (43 + 60 + 16 + 41) / 60  # get total hours in a single expression
total_hr

2.6666666666666665

<hr></hr>
<div class="alert alert-success alertsuccess" style="margin-top: 0px">
<h4> [Tip] Variables in Python </h4>
<p></p>
As you just learned, you can use **variables** to store values for repeated use. Here are some more **characteristics of variables in Python**:
<li>variables store the output of a block of code </li>
<li>variables are typically assigned using **=** (as in **x = 1**) </li>
<p></p>
</div>
<hr></hr>

 <a id="#ref1a"></a>
<h2 align=center> Quiz on Simple Math and Variables in Python</h2>

#### 1) What is the value of x ? <p> x=3+2*2</p>

<div align="right">
<a href="#SM1" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="SM1" class="collapse">
```
7
```
</div>


#### 2) What is the value of y ? <p> y=(3+2)*2</p>

<div align="right">
<a href="#SM2" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="SM2" class="collapse">
```
10
```
</div>


#### 3) What is the value of z ? <p> z=x+y</p>

<div align="right">
<a href="#SM3" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="SM3" class="collapse">
```
17
```
</div>

### <hr></hr>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
<h4> [Tip] # Comments</h4>  

Did you notice the **comment** after the **total_hr = (43 + 60 + 16 + 41) / 60 ** above? Comments are very useful in describing your code. You can create your own comments by using the **#** symbol and writing your comment after it. Python will interpret it as a comment, not as code.

<p></p>
</div>
<hr></hr>

<a id="ref10"></a>
<h2 align=center>Strings in Python</h2>

Python isn't just about numbers -- we can also have strings too. For example:

In [12]:
album = "Thriller"
album

'Thriller'

In Python, you can identify **Strings** when they are wrapped with **matching double (") or single (') quotes**.

Let's create a **String array** for the following **genres**:
- Punk
- Big Band
- Country
- Dance
- Folk
- Funk

In [13]:
genres = ["Punk", "Big Band", "Country", "Dance", "Folk", "Funk"] 
genres

['Punk', 'Big Band', 'Country', 'Dance', 'Folk', 'Funk']

<a id="ref11"></a>
<h2 align=center> Quiz on Strings </h2>

#### What is the value of the variable A after the following code is executed? 
`A="one"`

 <div align="right">
<a href="#String1" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="String1" class="collapse">
```
"one"
```
</div>


#### What is the value of the variable B after the following code is executed?
`B="2"`

 <div align="right">
<a href="#String2" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="String2" class="collapse">
```
"2"
```
</div>


#### What is the value of the variable C after the following code is executed?
`C=A+B`

 <div align="right">
<a href="#String3" class="btn btn-default" data-toggle="collapse">Click here for the solution</a>

</div>
<div id="String3" class="collapse">
```
"one2"
```
</div>

<hr></hr>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
<h4> [Tip] Saving the Notebook </h4>  

Your notebook saves automatically every two minutes. You can manually save by going to **File** > **Save and Checkpoint**. You can come back to this notebook anytime by clicking this notebook under the "**Recent Notebooks**" list on the right-hand side.   


</div>
<hr></hr>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
<h4> [Tip] Notebook Features </h4>  
  
Did you know there are other **notebook options**? Click on the **>** symbol to the left of the notebook:

<img src =https://ibm.box.com/shared/static/otu40m0kkzz5hropxah1nnzd2j01itom.png width = 35%>


<p></p>

</div>
<hr></hr>

### About the Authors:  

V2) This note book was updated by [Joseph Santarcangelo](https://www.linkedin.com/in/joseph-s-50398b136/) 

V1) Hi! It's [Agatha Colangelo](https://www.linkedin.com/in/agathacolangelo) and [Marta Aghili](https://ca.linkedin.com/in/marta-aghili-2b184b71), the authors of this notebook. We hope you found Python easy to learn! There's lots more to learn about Python but you're well on your way. Feel free to connect with us if you have any questions.

<hr>
Copyright &copy; 2016 [Big Data University](https://bigdatauniversity.com/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).​